### PR TITLE
Adds TypeVar with ForwardRef support

### LINKED ISSUE #2536: Incorrect `TypeVar` + `ForwardRef` resolution
Here's a source code that produces an error:

```python
from hypothesis import strategies as st
from typing import TypeVar

_T = TypeVar('_T', bound='Test')

def example(t: _T) -> None:
    return t.arg

class Test(object):
    def __init__(self, arg: int) -> None:
        self.arg = arg

print(st.builds(example).example())
```

It generates this error:

```
» python ex.py
/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py:268: NonInteractiveExampleWarning: The `.example()` method is good for exploring strategies, but should only be used interactively.  We recommend using `@given` for tests - it performs better, saves and replays failures to avoid flakiness, and reports minimal examples. (strategy: builds(example))
  warnings.warn(
Traceback (most recent call last):
  File "ex.py", line 13, in <module>
    print(st.builds(example).example())
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py", line 314, in example
    example_generating_inner_function()
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py", line 302, in example_generating_inner_function
    @settings(
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/core.py", line 1001, in wrapped_test
    processed_args = process_arguments_to_given(
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/core.py", line 420, in process_arguments_to_given
    search_strategy.validate()
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py", line 377, in validate
    self.do_validate()
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/collections.py", line 39, in do_validate
    s.validate()
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py", line 377, in validate
    self.do_validate()
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py", line 638, in do_validate
    self.mapped_strategy.validate()
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py", line 377, in validate
    self.do_validate()
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/lazy.py", line 118, in do_validate
    w.validate()
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py", line 377, in validate
    self.do_validate()
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py", line 638, in do_validate
    self.mapped_strategy.validate()
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py", line 377, in validate
    self.do_validate()
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/collections.py", line 39, in do_validate
    s.validate()
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py", line 377, in validate
    self.do_validate()
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/lazy.py", line 118, in do_validate
    w.validate()
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py", line 377, in validate
    self.do_validate()
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py", line 638, in do_validate
    self.mapped_strategy.validate()
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py", line 377, in validate
    self.do_validate()
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/lazy.py", line 118, in do_validate
    w.validate()
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py", line 377, in validate
    self.do_validate()
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/collections.py", line 39, in do_validate
    s.validate()
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py", line 377, in validate
    self.do_validate()
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py", line 638, in do_validate
    self.mapped_strategy.validate()
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py", line 377, in validate
    self.do_validate()
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/collections.py", line 39, in do_validate
    s.validate()
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py", line 377, in validate
    self.do_validate()
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/lazy.py", line 118, in do_validate
    w.validate()
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py", line 378, in validate
    self.is_empty
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py", line 125, in accept
    recur(self)
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py", line 121, in recur
    mapping[strat] = getattr(strat, calculation)(recur)
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/deferred.py", line 80, in calc_is_empty
    return recur(self.wrapped_strategy)
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/deferred.py", line 43, in wrapped_strategy
    result = self.__definition()
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/core.py", line 1385, in <lambda>
    lambda thing: deferred(lambda: _from_type(thing)),
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/core.py", line 1473, in _from_type
    return types.from_typing_type(thing)
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/types.py", line 152, in from_typing_type
    strat = unwrap_strategies(st.from_type(thing.__bound__))
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/lazy.py", line 45, in unwrap_strategies
    result = unwrap_strategies(s.wrapped_strategy)
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/lazy.py", line 45, in unwrap_strategies
    result = unwrap_strategies(s.wrapped_strategy)
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/deferred.py", line 43, in wrapped_strategy
    result = self.__definition()
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/core.py", line 1385, in <lambda>
    lambda thing: deferred(lambda: _from_type(thing)),
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/core.py", line 1473, in _from_type
    return types.from_typing_type(thing)
  File "/Users/sobolev/Documents/github/returns/.venv/lib/python3.8/site-packages/hypothesis/strategies/_internal/types.py", line 229, in from_typing_type
    raise ResolutionFailed(
hypothesis.errors.ResolutionFailed: Could not resolve ForwardRef('Test') to a strategy; consider using register_type_strategy
```

I guess that this happens due to lack of `ForwardRef` handling here: https://github.com/HypothesisWorks/hypothesis/blob/7d6d048ad5df1d01a4ddc90f1b8f2bb45fe2f0de/hypothesis-python/src/hypothesis/strategies/_internal/types.py#L104

### PR DESCRIPTION
**Pull request by sobolevn**
Closes #2536 

There's a lot of thing to test before actually merging this thing 😄

### CODE REVIEW

**Comment by sobolevn** [comment]
We won't be able to resolve things that are hidden under `if TYPE_CHECKING` and that's quite a popular pattern. 
Maybe we should also try to resolve `bound=` from registered type strategies?

**Comment by Zac-HD** [comment]
- Needs a `hypothesis-python/RELEASE.rst`, for a `minor` release just like last time.
- Again, I'm OK with just skipping tests on Python < 3.7 if that's easier to start with.

> We won't be able to resolve things that are hidden under `if TYPE_CHECKING` and that's quite a popular pattern.

I think this is just too bad; `if TYPE_CHECKING:` is just incompatible with resolving the references.

Can you give me an example of a case with a bound that you propose to resolve?  This sounds OK but I'm not sure I understand what you want to do.

**Comment by sobolevn** [comment]
@Zac-HD I was refererring to this example with `TYPE_CHECKING`: https://github.com/HypothesisWorks/hypothesis/pull/2550/files#diff-db60ee6b6bd001ff0051be1f3a2df16aR58-R67

File: hypothesis-python/src/hypothesis/strategies/_internal/types.py [hunk 1]
```diff
@@ -131,6 +132,33 @@ def is_generic_type(type_):
...
+        raise InvalidArgument(
+            "Looks like you are using python3.6 or any of the previous versions "
+            "together with a TypeVar bound to a ForwardRef. "
+            "It is not supported. Upgrading to python3.7+ will solve this problem."
+        )
+
+    try:
+        module = importlib.import_module(thing.__module__)
+        return typing._eval_type(bound, vars(module), None)
+    except (ImportError, AttributeError, NameError):
```
[the remark this answers is not in the document]
Comment (sobolevn) [reply hypothesis-python/src/hypothesis/strategies/_internal/types.py]: This didn't work:

```
  File "/Users/sobolev/.pyenv/versions/3.6.8/lib/python3.6/typing.py", line 350, in _eval_type
    return t._eval_type(globalns, localns)
  File "/Users/sobolev/.pyenv/versions/3.6.8/lib/python3.6/typing.py", line 245, in _eval_type
    eval(self.__forward_code__, globalns, localns),
TypeError: locals must be a mapping
```

[the remark this answers is not in the document]
Comment (sobolevn) [reply hypothesis-python/src/hypothesis/strategies/_internal/types.py]: This does work:

```python
return typing._eval_type(bound, vars(sys.modules[thing.__module__]), None)
```

**Review by Zac-HD — APPROVED** [review]
:tada: :shipit:

**Comment by Zac-HD** [comment]
Ideas for next steps:

- I've mentioned the `mock.patch` issue (2231) before, or pick anything that looks interesting
- it would also be great to have more people involved in code review ([our guide](https://github.com/HypothesisWorks/hypothesis/blob/master/guides/review.rst)) - comments *or questions* about anything from high level design to fiddly implementation are welcome

**Comment by sobolevn** [comment]
Thanks a lot for your help! 🥇

**Pull request merged**